src/backend/distributed/worker/worker_data_fetch_protocol.c

```diff
@@ -662,17 +664,30 @@ worker_append_table_to_shard(PG_FUNCTION_ARGS)
 	/* make sure we are allowed to execute the COPY command */
 	CheckCopyPermissions(localCopyCommand);
 
-	/* need superuser to copy from files */
-	GetUserIdAndSecContext(&savedUserId, &savedSecurityContext);
-	SetUserIdAndSecContext(CitusExtensionOwner(), SECURITY_LOCAL_USERID_CHANGE);
+	Relation shardRelation = table_openrv(localCopyCommand->relation, RowExclusiveLock);
+
+	/* mimic check from copy.c */
+	if (XactReadOnly && !shardRelation->rd_islocaltemp)
+	{
+		PreventCommandIfReadOnly("COPY FROM");
+	}
+
+	CopyFromState copyState = BeginCopyFrom_compat(NULL,
```

Comment:
We might need to pass `ParseState`, it seems that's the reason for the crash in PG14:
```C
#include "parser/parse_relation.h"
	ParseState *pState = make_parsestate(NULL);
	(void) addRangeTableEntryForRelation(pState, shardRelation, AccessShareLock,
										 NULL, false, false);
	CopyFromState copyState = BeginCopyFrom_compat(pState,...
```

Reply:
Thanks!